File: 93-restore-ip-addresses/93-restore-ip-addresses.py
class Solution:
    def restoreIpAddresses(self, s: str) -> List[str]:
        if len(s) < 4 or len(s) > 12:
            return []
        def combine(temp,root,nxt):
            n = len(nxt)
            for i in range(n):
                temp.append([root]+nxt[i])
                
        def generate(s,n):
            if n == 3:
                if len(s) > 0 and ((int(s) <= 255 and s[0] != '0') or s == '0'):
                    return True,[[s]]
                else:
                    return False,None
            temp = []
            for i in range(3):
                if (i+1) > len(s) or int(s[:(i+1)]) > 255 or (s[0] == '0' and i>0):
                    continue
                nxt = generate(s[i+1:],n+1)
                if nxt[0]:
                    combine(temp,s[:(i+1)],nxt[1])
            return True,temp
        res = generate(s,0)[1]
        final = []
        for elem in res:
            final.append('.'.join(elem))
        return final
            
                
            
        

File: 93-restore-ip-addresses/93-restore-ip-addresses.java
class Solution {
    StringBuilder temp = new StringBuilder();
    List<String> res = new LinkedList<>();
    int n = 0;
    public List<String> restoreIpAddresses(String s) {
        if(n == 4 && s.length() == 0){
            res.add(temp.toString());
            return res;
        }
        if(n>0)temp.append(".");
        for(int i = 0; i < Math.min(s.length(),3); i++){
            String cur = s.substring(0,i+1);
            if(cur.charAt(0) == '0' && cur.length() > 1)continue;
        //    System.out.println(cur);
            int curVal = Integer.parseInt(cur);
            if(curVal > 255)break;
            n += 1;
            temp.append(cur);
            restoreIpAddresses(s.substring(i+1));
            n -= 1;
            temp.setLength(temp.length() - i-1);
        }
        if(n > 0)temp.setLength(temp.length() - 1);
        return res;
        
    }
}